---
title: "Combined Samples - 03 DEG Analysis"
author:
  - name: "Emir Turkes"
date: '`r strftime(Sys.time(), format = "%B %d, %Y")`'
link-citations: true
output:
  html_document:
    code_folding: hide
    number_sections: true
    theme: lumen
    highlight: haddock
    toc: true
    toc_depth: 3
    toc_float:
      collapsed: false
      smooth_scroll: false
knit: (function(inputFile, encoding) {
  rmarkdown::render(
    inputFile, encoding = encoding,
    output_file = file.path("..", "..", "results", unlist(strsplit(getwd(), "/"))[6], "comb_03_DEG.html")
  )})
---

<style type="text/css">
body {font-size: 16px;}
h1.title {font-size: 35px;}
h1 {font-size: 24px;}
h2 {font-size: 22px;}
h3 {font-size: 20px;}
.toc-content {padding-left: 0px; padding-right: 0px;}
div.tocify {width: 100%;}
.tocify-subheader .tocify-item {font-size: 0.95em; padding-left: 25px; text-indent: 0;}
.tocify-subheader .tocify-subheader .tocify-item {font-size: 0.95em; padding-left: 35px; text-indent: 0;}
div.main-container {max-width: none; width: 100%;}
</style>

*This file is a part of [tau-mutant-omics-analysis](https://github.com/eturkes/tau-mutant-omics-analysis).*

The table of contents in the top left is clickable and can be used to quickly navigate the document.
To toggle the visibility of code, use the `CODE` toggles at the top right of chunks.
The toggle at the start of the document controls the visibility of all chunks.

```{r}
#    This file is part of tau-mutant-omics-analysis.
#    Copyright (C) 2023  Emir Turkes, Naoto Watamura, UK DRI at UCL
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#    Emir Turkes can be contacted at emir.turkes@eturkes.com

# Load required packages, suppressing startup messages.
# -----------------------------------------------------
library(conflicted)
conflicts_prefer(base::intersect, DT::JS, .quiet = TRUE)
packages <- c(
  "Seurat", "SingleCellExperiment", "scuttle", "parallelly", "ggplot2", "dplyr", "ggrepel", "viridis", "DT", "patchwork"
)
invisible(suppressPackageStartupMessages(lapply(packages, FUN = library, character.only = TRUE)))
# -----------------------------------------------------

# Define global settings.
# -----------------------
knitr::opts_chunk$set(fig.width = 20, fig.height = 12, dpi = 150)
# -----------------------

# Define functions.
# -----------------
source(file.path("..", "utils.R"))
# -----------------

# Useful variables.
# -----------------
analysis_no <- 3
data_name <- unlist(strsplit(getwd(), "/"))[6] # Name of dataset derived from path.

cache_dir <- file.path("..", "..", "cache", data_name, paste0("0", analysis_no)) # Store intermediate data objects here.
if (!dir.exists(cache_dir)) {
  dir.create(cache_dir, recursive = TRUE)
}

protocol <- c("mouse", "droplet", "single-nuc", "umis") # See `cluster_pipeline` in `utils.R`.
vars_to_regress <- NULL # See `cluster_pipeline` in `utils.R`.
parallel_override <- 0 # See `parallel_plan` in `utils.R`.
# -----------------

# Metadata to plot after dimensionality reduction and clustering.
# Values in list can include "no_legend and/or "no_label" to exclude those.
# -------------------------------------------------------------------------
metadata_to_plot <- vector("list", length = 4)
names(metadata_to_plot) <- c("seurat_clusters", "seurat_clusters_introns", "sample", "doublets")
metadata_to_plot[[3]] <- "no_label"
metadata_to_plot[[4]] <- "no_label"
# -------------------------------------------------------------------------
```

# Prep

```{r, fig.width = 20, fig.height = 14}
rds <- file.path(cache_dir, "merged_seurat.rds")
if (file.exists(rds)) {
  seurat <- readRDS(rds)

  seurat_introns <- readRDS(file.path(cache_dir, "..", "02", "seurat.rds"))
  seurat_introns <- seurat_introns[ , colnames(seurat_introns) %in% colnames(seurat)]

} else {

  file_list <- list.files(file.path(cache_dir, "..", ".."), recursive = TRUE)
  file_list <- file_list[grep("no-introns", file_list)]
  files <- vector("list", length(file_list))
  for (i in seq_along(files)) {
    files[[i]] <- readRDS(file.path(cache_dir, "..", "..", file_list[i]))
  }

  genes <- Reduce(
    intersect,
    list(
      rownames(files[[1]]), rownames(files[[2]]), rownames(files[[3]]),
      rownames(files[[4]]), rownames(files[[5]]), rownames(files[[6]])
    )
  )
  for (i in seq_along(files)) {
    DefaultAssay(files[[i]]) <- "RNA"
    files[[i]][["SCT"]] <- NULL
  }

  seurat <- merge(files[[1]], c(files[[2]], files[[3]], files[[4]], files[[5]], files[[6]]))
  seurat@meta.data[grep("SCT", colnames(seurat@meta.data))] <- NULL

  seurat$sample <- c(
    rep(sub("\\/.*", "", file_list[1]), ncol(files[[1]])),
    rep(sub("\\/.*", "", file_list[2]), ncol(files[[2]])),
    rep(sub("\\/.*", "", file_list[3]), ncol(files[[3]])),
    rep(sub("\\/.*", "", file_list[4]), ncol(files[[4]])),
    rep(sub("\\/.*", "", file_list[5]), ncol(files[[5]])),
    rep(sub("\\/.*", "", file_list[6]), ncol(files[[6]]))
  )

  ribo <- grep("^RP[SL][[:digit:]]", rownames(seurat), ignore.case = TRUE)
  sce <- as.SingleCellExperiment(seurat)
  sce <- addPerCellQC(sce, list(ribo = ribo))
  seurat$subsets_ribo_percent <- sce$subsets_ribo_percent
  rm(sce)

  seurat_introns <- readRDS(file.path(cache_dir, "..", "02", "seurat.rds"))
  seurat_introns <- seurat_introns[ , colnames(seurat_introns) %in% colnames(seurat)]
  seurat <- seurat[ , colnames(seurat) %in% colnames(seurat_introns)]
  seurat$seurat_clusters_introns <- seurat_introns$seurat_clusters
  seurat$umap1_introns <- seurat_introns$umap1
  seurat$umap2_introns <- seurat_introns$umap2
  seurat$predicted.id <- seurat_introns$predicted.id
  seurat$doublets <- seurat_introns$doublets

  saveRDS(seurat, file = rds)
}

sub_name <- "processed_merged"
seurat <- cluster_pipeline(
  seurat, cache_dir = cache_dir, sub_name = sub_name, protocol = protocol, vars_to_regress = vars_to_regress,
  parallel_override = parallel_override, cc = FALSE, res_divider = 1000000, conserve_memory = TRUE
)
seurat

seurat$seurat_clusters_introns_sample <- paste(seurat$seurat_clusters_introns, seurat$sample, sep = "_")
seurat_introns$seurat_clusters_sample <- paste(seurat_introns$seurat_clusters, seurat$sample, sep = "_")
```

# Exons Only

## NLGF P301S+3 vs. NLGF MAPT KI

```{r}
ident1 <- "NLGF_P301S+3"
ident2 <- "NLGF_MAPT_KI"

seurat@active.ident <- factor(seurat$sample)
markers <- FindMarkers(seurat, ident.1 = ident1, ident.2 = ident2, only.pos = FALSE, verbose = FALSE)
markers <- markers[markers$p_val_adj < 0.05, ]
datatable_download_exp(markers)

seurat_sub <- subset(seurat, idents = c(ident1, ident2))
VlnPlot(seurat_sub, features = rownames(markers)[1:6], group.by = "sample") + plot_annotation("Exons Only")

seurat@active.ident <- factor(seurat$seurat_clusters_introns_sample)
```

### Oligodendrocytes

```{r}
cluster_num <- 2

markers <- FindMarkers(
  seurat, ident.1 = paste(cluster_num, ident1, sep = "_"), ident.2 = paste(cluster_num, ident2, sep = "_"),
  only.pos = FALSE, verbose = FALSE
)
markers <- markers[markers$p_val_adj < 0.05, ]
datatable_download_exp(markers)

seurat_sub <- subset(seurat, idents = c(paste(cluster_num, ident1, sep = "_"), paste(cluster_num, ident2, sep = "_")))
seurat_sub$seurat_clusters_introns_sample_names <- sub(
  cluster_num, "Oligodendrocytes", seurat_sub$seurat_clusters_introns_sample
)
VlnPlot(seurat_sub, features = rownames(markers)[1:6], group.by = "seurat_clusters_introns_sample_names") +
  plot_annotation("Exons Only")
```

### Microglia

```{r}
cluster_num <- 11

markers <- FindMarkers(
  seurat, ident.1 = paste(cluster_num, ident1, sep = "_"), ident.2 = paste(cluster_num, ident2, sep = "_"),
  only.pos = FALSE, verbose = FALSE
)
markers <- markers[markers$p_val_adj < 0.05, ]
datatable_download_exp(markers)

seurat_sub <- subset(seurat, idents = c(paste(cluster_num, ident1, sep = "_"), paste(cluster_num, ident2, sep = "_")))
seurat_sub$seurat_clusters_introns_sample_names <- sub(
  cluster_num, "Microglia", seurat_sub$seurat_clusters_introns_sample
)
VlnPlot(seurat_sub, features = rownames(markers)[1:6], group.by = "seurat_clusters_introns_sample_names") +
  plot_annotation("Exons Only")
```

### Astrocytes

```{r}
cluster_num <- 1

markers <- FindMarkers(
  seurat, ident.1 = paste(cluster_num, ident1, sep = "_"), ident.2 = paste(cluster_num, ident2, sep = "_"),
  only.pos = FALSE, verbose = FALSE
)
markers <- markers[markers$p_val_adj < 0.05, ]
datatable_download_exp(markers)

seurat_sub <- subset(seurat, idents = c(paste(cluster_num, ident1, sep = "_"), paste(cluster_num, ident2, sep = "_")))
seurat_sub$seurat_clusters_introns_sample_names <- sub(
  cluster_num, "Astrocytes", seurat_sub$seurat_clusters_introns_sample
)
VlnPlot(seurat_sub, features = rownames(markers)[1:6], group.by = "seurat_clusters_introns_sample_names") +
  plot_annotation("Exons Only")
```

### Cortical Glutamatergic

```{r}
cluster_num <- 0

markers <- FindMarkers(
  seurat, ident.1 = paste(cluster_num, ident1, sep = "_"), ident.2 = paste(cluster_num, ident2, sep = "_"),
  only.pos = FALSE, verbose = FALSE
)
markers <- markers[markers$p_val_adj < 0.05, ]
datatable_download_exp(markers)

seurat_sub <- subset(seurat, idents = c(paste(cluster_num, ident1, sep = "_"), paste(cluster_num, ident2, sep = "_")))
seurat_sub$seurat_clusters_introns_sample_names <- sub(
  cluster_num, "Ctx_Glutamatergic", seurat_sub$seurat_clusters_introns_sample
)
VlnPlot(seurat_sub, features = rownames(markers)[1:6], group.by = "seurat_clusters_introns_sample_names") +
  plot_annotation("Exons Only")
```

### GABAergic

```{r}
cluster_num <- 3

markers <- FindMarkers(
  seurat, ident.1 = paste(cluster_num, ident1, sep = "_"), ident.2 = paste(cluster_num, ident2, sep = "_"),
  only.pos = FALSE, verbose = FALSE
)
markers <- markers[markers$p_val_adj < 0.05, ]
datatable_download_exp(markers)

seurat_sub <- subset(seurat, idents = c(paste(cluster_num, ident1, sep = "_"), paste(cluster_num, ident2, sep = "_")))
seurat_sub$seurat_clusters_introns_sample_names <- sub(
  cluster_num, "GABAergic", seurat_sub$seurat_clusters_introns_sample
)
VlnPlot(seurat_sub, features = rownames(markers)[1:6], group.by = "seurat_clusters_introns_sample_names") +
  plot_annotation("Exons Only")
```

## NLGF MAPT KI vs. MAPT KI

```{r}
ident1 <- "NLGF_MAPT_KI"
ident2 <- "MAPT_KI"

seurat@active.ident <- factor(seurat$sample)
markers <- FindMarkers(seurat, ident.1 = ident1, ident.2 = ident2, only.pos = FALSE, verbose = FALSE)
markers <- markers[markers$p_val_adj < 0.05, ]
datatable_download_exp(markers)

seurat_sub <- subset(seurat, idents = c(ident1, ident2))
VlnPlot(seurat_sub, features = rownames(markers)[1:6], group.by = "sample") + plot_annotation("Exons Only")

seurat@active.ident <- factor(seurat$seurat_clusters_introns_sample)
```

### Oligodendrocytes

```{r}
cluster_num <- 2

markers <- FindMarkers(
  seurat, ident.1 = paste(cluster_num, ident1, sep = "_"), ident.2 = paste(cluster_num, ident2, sep = "_"),
  only.pos = FALSE, verbose = FALSE
)
markers <- markers[markers$p_val_adj < 0.05, ]
datatable_download_exp(markers)

seurat_sub <- subset(seurat, idents = c(paste(cluster_num, ident1, sep = "_"), paste(cluster_num, ident2, sep = "_")))
seurat_sub$seurat_clusters_introns_sample_names <- sub(
  cluster_num, "Oligodendrocytes", seurat_sub$seurat_clusters_introns_sample
)
VlnPlot(seurat_sub, features = rownames(markers)[1:6], group.by = "seurat_clusters_introns_sample_names") +
  plot_annotation("Exons Only")
```

### Microglia

```{r}
cluster_num <- 11

markers <- FindMarkers(
  seurat, ident.1 = paste(cluster_num, ident1, sep = "_"), ident.2 = paste(cluster_num, ident2, sep = "_"),
  only.pos = FALSE, verbose = FALSE
)
markers <- markers[markers$p_val_adj < 0.05, ]
datatable_download_exp(markers)

seurat_sub <- subset(seurat, idents = c(paste(cluster_num, ident1, sep = "_"), paste(cluster_num, ident2, sep = "_")))
seurat_sub$seurat_clusters_introns_sample_names <- sub(
  cluster_num, "Microglia", seurat_sub$seurat_clusters_introns_sample
)
VlnPlot(seurat_sub, features = rownames(markers)[1:6], group.by = "seurat_clusters_introns_sample_names") +
  plot_annotation("Exons Only")
```

### Astrocytes

```{r}
cluster_num <- 1

markers <- FindMarkers(
  seurat, ident.1 = paste(cluster_num, ident1, sep = "_"), ident.2 = paste(cluster_num, ident2, sep = "_"),
  only.pos = FALSE, verbose = FALSE
)
markers <- markers[markers$p_val_adj < 0.05, ]
datatable_download_exp(markers)

seurat_sub <- subset(seurat, idents = c(paste(cluster_num, ident1, sep = "_"), paste(cluster_num, ident2, sep = "_")))
seurat_sub$seurat_clusters_introns_sample_names <- sub(
  cluster_num, "Astrocytes", seurat_sub$seurat_clusters_introns_sample
)
VlnPlot(seurat_sub, features = rownames(markers)[1:6], group.by = "seurat_clusters_introns_sample_names") +
  plot_annotation("Exons Only")
```

### Cortical Glutamatergic

```{r}
cluster_num <- 0

markers <- FindMarkers(
  seurat, ident.1 = paste(cluster_num, ident1, sep = "_"), ident.2 = paste(cluster_num, ident2, sep = "_"),
  only.pos = FALSE, verbose = FALSE
)
markers <- markers[markers$p_val_adj < 0.05, ]
datatable_download_exp(markers)

seurat_sub <- subset(seurat, idents = c(paste(cluster_num, ident1, sep = "_"), paste(cluster_num, ident2, sep = "_")))
seurat_sub$seurat_clusters_introns_sample_names <- sub(
  cluster_num, "Ctx_Glutamatergic", seurat_sub$seurat_clusters_introns_sample
)
VlnPlot(seurat_sub, features = rownames(markers)[1:6], group.by = "seurat_clusters_introns_sample_names") +
  plot_annotation("Exons Only")
```

### GABAergic

```{r}
cluster_num <- 3

markers <- FindMarkers(
  seurat, ident.1 = paste(cluster_num, ident1, sep = "_"), ident.2 = paste(cluster_num, ident2, sep = "_"),
  only.pos = FALSE, verbose = FALSE
)
markers <- markers[markers$p_val_adj < 0.05, ]
datatable_download_exp(markers)

seurat_sub <- subset(seurat, idents = c(paste(cluster_num, ident1, sep = "_"), paste(cluster_num, ident2, sep = "_")))
seurat_sub$seurat_clusters_introns_sample_names <- sub(
  cluster_num, "GABAergic", seurat_sub$seurat_clusters_introns_sample
)
VlnPlot(seurat_sub, features = rownames(markers)[1:6], group.by = "seurat_clusters_introns_sample_names") +
  plot_annotation("Exons Only")
```

## NLGF P301S+3 vs. MAPT KI

```{r}
ident1 <- "NLGF_P301S+3"
ident2 <- "MAPT_KI"

seurat@active.ident <- factor(seurat$sample)
markers <- FindMarkers(seurat, ident.1 = ident1, ident.2 = ident2, only.pos = FALSE, verbose = FALSE)
markers <- markers[markers$p_val_adj < 0.05, ]
datatable_download_exp(markers)

seurat_sub <- subset(seurat, idents = c(ident1, ident2))
VlnPlot(seurat_sub, features = rownames(markers)[1:6], group.by = "sample") + plot_annotation("Exons Only")

seurat@active.ident <- factor(seurat$seurat_clusters_introns_sample)
```

### Oligodendrocytes

```{r}
cluster_num <- 2

markers <- FindMarkers(
  seurat, ident.1 = paste(cluster_num, ident1, sep = "_"), ident.2 = paste(cluster_num, ident2, sep = "_"),
  only.pos = FALSE, verbose = FALSE
)
markers <- markers[markers$p_val_adj < 0.05, ]
datatable_download_exp(markers)

seurat_sub <- subset(seurat, idents = c(paste(cluster_num, ident1, sep = "_"), paste(cluster_num, ident2, sep = "_")))
seurat_sub$seurat_clusters_introns_sample_names <- sub(
  cluster_num, "Oligodendrocytes", seurat_sub$seurat_clusters_introns_sample
)
VlnPlot(seurat_sub, features = rownames(markers)[1:6], group.by = "seurat_clusters_introns_sample_names") +
  plot_annotation("Exons Only")
```

### Microglia

```{r}
cluster_num <- 11

markers <- FindMarkers(
  seurat, ident.1 = paste(cluster_num, ident1, sep = "_"), ident.2 = paste(cluster_num, ident2, sep = "_"),
  only.pos = FALSE, verbose = FALSE
)
markers <- markers[markers$p_val_adj < 0.05, ]
datatable_download_exp(markers)

seurat_sub <- subset(seurat, idents = c(paste(cluster_num, ident1, sep = "_"), paste(cluster_num, ident2, sep = "_")))
seurat_sub$seurat_clusters_introns_sample_names <- sub(
  cluster_num, "Microglia", seurat_sub$seurat_clusters_introns_sample
)
VlnPlot(seurat_sub, features = rownames(markers)[1:6], group.by = "seurat_clusters_introns_sample_names") +
  plot_annotation("Exons Only")
```

### Astrocytes

```{r}
cluster_num <- 1

markers <- FindMarkers(
  seurat, ident.1 = paste(cluster_num, ident1, sep = "_"), ident.2 = paste(cluster_num, ident2, sep = "_"),
  only.pos = FALSE, verbose = FALSE
)
markers <- markers[markers$p_val_adj < 0.05, ]
datatable_download_exp(markers)

seurat_sub <- subset(seurat, idents = c(paste(cluster_num, ident1, sep = "_"), paste(cluster_num, ident2, sep = "_")))
seurat_sub$seurat_clusters_introns_sample_names <- sub(
  cluster_num, "Astrocytes", seurat_sub$seurat_clusters_introns_sample
)
VlnPlot(seurat_sub, features = rownames(markers)[1:6], group.by = "seurat_clusters_introns_sample_names") +
  plot_annotation("Exons Only")
```

### Cortical Glutamatergic

```{r}
cluster_num <- 0

markers <- FindMarkers(
  seurat, ident.1 = paste(cluster_num, ident1, sep = "_"), ident.2 = paste(cluster_num, ident2, sep = "_"),
  only.pos = FALSE, verbose = FALSE
)
markers <- markers[markers$p_val_adj < 0.05, ]
datatable_download_exp(markers)

seurat_sub <- subset(seurat, idents = c(paste(cluster_num, ident1, sep = "_"), paste(cluster_num, ident2, sep = "_")))
seurat_sub$seurat_clusters_introns_sample_names <- sub(
  cluster_num, "Ctx_Glutamatergic", seurat_sub$seurat_clusters_introns_sample
)
VlnPlot(seurat_sub, features = rownames(markers)[1:6], group.by = "seurat_clusters_introns_sample_names") +
  plot_annotation("Exons Only")
```

### GABAergic

```{r}
cluster_num <- 3

markers <- FindMarkers(
  seurat, ident.1 = paste(cluster_num, ident1, sep = "_"), ident.2 = paste(cluster_num, ident2, sep = "_"),
  only.pos = FALSE, verbose = FALSE
)
markers <- markers[markers$p_val_adj < 0.05, ]
datatable_download_exp(markers)

seurat_sub <- subset(seurat, idents = c(paste(cluster_num, ident1, sep = "_"), paste(cluster_num, ident2, sep = "_")))
seurat_sub$seurat_clusters_introns_sample_names <- sub(
  cluster_num, "GABAergic", seurat_sub$seurat_clusters_introns_sample
)
VlnPlot(seurat_sub, features = rownames(markers)[1:6], group.by = "seurat_clusters_introns_sample_names") +
  plot_annotation("Exons Only")
```

# Exons + Introns

## NLGF P301S+3 vs. NLGF MAPT KI

```{r}
ident1 <- "NLGF_P301S+3"
ident2 <- "NLGF_MAPT_KI"

seurat_introns@active.ident <- factor(seurat_introns$sample)
markers <- FindMarkers(seurat_introns, ident.1 = ident1, ident.2 = ident2, only.pos = FALSE, verbose = FALSE)
markers <- markers[markers$p_val_adj < 0.05, ]
datatable_download_exp(markers)

seurat_sub <- subset(seurat_introns, idents = c(ident1, ident2))
VlnPlot(seurat_sub, features = rownames(markers)[1:6], group.by = "sample") + plot_annotation("Exons + Introns")

seurat_introns@active.ident <- factor(seurat_introns$seurat_clusters_sample)
```

### Oligodendrocytes

```{r}
cluster_num <- 2

markers <- FindMarkers(
  seurat_introns, ident.1 = paste(cluster_num, ident1, sep = "_"), ident.2 = paste(cluster_num, ident2, sep = "_"),
  only.pos = FALSE, verbose = FALSE
)
markers <- markers[markers$p_val_adj < 0.05, ]
datatable_download_exp(markers)

seurat_sub <- subset(
  seurat_introns, idents = c(paste(cluster_num, ident1, sep = "_"), paste(cluster_num, ident2, sep = "_"))
)
seurat_sub$seurat_clusters_sample_names <- sub(cluster_num, "Oligodendrocytes", seurat_sub$seurat_clusters_sample)
VlnPlot(seurat_sub, features = rownames(markers)[1:6], group.by = "seurat_clusters_sample_names") +
  plot_annotation("Exons + Introns")
```

### Microglia

```{r}
cluster_num <- 11

markers <- FindMarkers(
  seurat_introns, ident.1 = paste(cluster_num, ident1, sep = "_"), ident.2 = paste(cluster_num, ident2, sep = "_"),
  only.pos = FALSE, verbose = FALSE
)
markers <- markers[markers$p_val_adj < 0.05, ]
datatable_download_exp(markers)

seurat_sub <- subset(
  seurat_introns, idents = c(paste(cluster_num, ident1, sep = "_"), paste(cluster_num, ident2, sep = "_"))
)
seurat_sub$seurat_clusters_sample_names <- sub(cluster_num, "Microglia", seurat_sub$seurat_clusters_sample)
VlnPlot(seurat_sub, features = rownames(markers)[1:6], group.by = "seurat_clusters_sample_names") +
  plot_annotation("Exons + Introns")
```

### Astrocytes

```{r}
cluster_num <- 1

markers <- FindMarkers(
  seurat_introns, ident.1 = paste(cluster_num, ident1, sep = "_"), ident.2 = paste(cluster_num, ident2, sep = "_"),
  only.pos = FALSE, verbose = FALSE
)
markers <- markers[markers$p_val_adj < 0.05, ]
datatable_download_exp(markers)

seurat_sub <- subset(
  seurat_introns, idents = c(paste(cluster_num, ident1, sep = "_"), paste(cluster_num, ident2, sep = "_"))
)
seurat_sub$seurat_clusters_sample_names <- sub(cluster_num, "Astrocytes", seurat_sub$seurat_clusters_sample)
VlnPlot(seurat_sub, features = rownames(markers)[1:6], group.by = "seurat_clusters_sample_names") +
  plot_annotation("Exons + Introns")
```

### Cortical Glutamatergic

```{r}
cluster_num <- 0

markers <- FindMarkers(
  seurat_introns, ident.1 = paste(cluster_num, ident1, sep = "_"), ident.2 = paste(cluster_num, ident2, sep = "_"),
  only.pos = FALSE, verbose = FALSE
)
markers <- markers[markers$p_val_adj < 0.05, ]
datatable_download_exp(markers)

seurat_sub <- subset(
  seurat_introns, idents = c(paste(cluster_num, ident1, sep = "_"), paste(cluster_num, ident2, sep = "_"))
)
seurat_sub$seurat_clusters_sample_names <- sub(cluster_num, "Ctx_Glutamatergic", seurat_sub$seurat_clusters_sample)
VlnPlot(seurat_sub, features = rownames(markers)[1:6], group.by = "seurat_clusters_sample_names") +
  plot_annotation("Exons + Introns")
```

### GABAergic

```{r}
cluster_num <- 3

markers <- FindMarkers(
  seurat_introns, ident.1 = paste(cluster_num, ident1, sep = "_"), ident.2 = paste(cluster_num, ident2, sep = "_"),
  only.pos = FALSE, verbose = FALSE
)
markers <- markers[markers$p_val_adj < 0.05, ]
datatable_download_exp(markers)

seurat_sub <- subset(
  seurat_introns, idents = c(paste(cluster_num, ident1, sep = "_"), paste(cluster_num, ident2, sep = "_"))
)
seurat_sub$seurat_clusters_sample_names <- sub(cluster_num, "GABAergic", seurat_sub$seurat_clusters_sample)
VlnPlot(seurat_sub, features = rownames(markers)[1:6], group.by = "seurat_clusters_sample_names") +
  plot_annotation("Exons + Introns")
```

## NLGF MAPT KI vs. MAPT KI

```{r}
ident1 <- "NLGF_MAPT_KI"
ident2 <- "MAPT_KI"

seurat_introns@active.ident <- factor(seurat_introns$sample)
markers <- FindMarkers(seurat_introns, ident.1 = ident1, ident.2 = ident2, only.pos = FALSE, verbose = FALSE)
markers <- markers[markers$p_val_adj < 0.05, ]
datatable_download_exp(markers)

seurat_sub <- subset(seurat_introns, idents = c(ident1, ident2))
VlnPlot(seurat_sub, features = rownames(markers)[1:6], group.by = "sample") + plot_annotation("Exons + Introns")

seurat_introns@active.ident <- factor(seurat_introns$seurat_clusters_sample)
```

### Oligodendrocytes

```{r}
cluster_num <- 2

markers <- FindMarkers(
  seurat_introns, ident.1 = paste(cluster_num, ident1, sep = "_"), ident.2 = paste(cluster_num, ident2, sep = "_"),
  only.pos = FALSE, verbose = FALSE
)
markers <- markers[markers$p_val_adj < 0.05, ]
datatable_download_exp(markers)

seurat_sub <- subset(
  seurat_introns, idents = c(paste(cluster_num, ident1, sep = "_"), paste(cluster_num, ident2, sep = "_"))
)
seurat_sub$seurat_clusters_sample_names <- sub(cluster_num, "Oligodendrocytes", seurat_sub$seurat_clusters_sample)
VlnPlot(seurat_sub, features = rownames(markers)[1:6], group.by = "seurat_clusters_sample_names") +
  plot_annotation("Exons + Introns")
```

### Microglia

```{r}
cluster_num <- 11

markers <- FindMarkers(
  seurat_introns, ident.1 = paste(cluster_num, ident1, sep = "_"), ident.2 = paste(cluster_num, ident2, sep = "_"),
  only.pos = FALSE, verbose = FALSE
)
markers <- markers[markers$p_val_adj < 0.05, ]
datatable_download_exp(markers)

seurat_sub <- subset(
  seurat_introns, idents = c(paste(cluster_num, ident1, sep = "_"), paste(cluster_num, ident2, sep = "_"))
)
seurat_sub$seurat_clusters_sample_names <- sub(cluster_num, "Microglia", seurat_sub$seurat_clusters_sample)
VlnPlot(seurat_sub, features = rownames(markers)[1:6], group.by = "seurat_clusters_sample_names") +
  plot_annotation("Exons + Introns")
```

### Astrocytes

```{r}
cluster_num <- 1

markers <- FindMarkers(
  seurat_introns, ident.1 = paste(cluster_num, ident1, sep = "_"), ident.2 = paste(cluster_num, ident2, sep = "_"),
  only.pos = FALSE, verbose = FALSE
)
markers <- markers[markers$p_val_adj < 0.05, ]
datatable_download_exp(markers)

seurat_sub <- subset(
  seurat_introns, idents = c(paste(cluster_num, ident1, sep = "_"), paste(cluster_num, ident2, sep = "_"))
)
seurat_sub$seurat_clusters_sample_names <- sub(cluster_num, "Astrocytes", seurat_sub$seurat_clusters_sample)
VlnPlot(seurat_sub, features = rownames(markers)[1:6], group.by = "seurat_clusters_sample_names") +
  plot_annotation("Exons + Introns")
```

### Cortical Glutamatergic

```{r}
cluster_num <- 0

markers <- FindMarkers(
  seurat_introns, ident.1 = paste(cluster_num, ident1, sep = "_"), ident.2 = paste(cluster_num, ident2, sep = "_"),
  only.pos = FALSE, verbose = FALSE
)
markers <- markers[markers$p_val_adj < 0.05, ]
datatable_download_exp(markers)

seurat_sub <- subset(
  seurat_introns, idents = c(paste(cluster_num, ident1, sep = "_"), paste(cluster_num, ident2, sep = "_"))
)
seurat_sub$seurat_clusters_sample_names <- sub(cluster_num, "Ctx_Glutamatergic", seurat_sub$seurat_clusters_sample)
VlnPlot(seurat_sub, features = rownames(markers)[1:6], group.by = "seurat_clusters_sample_names") +
  plot_annotation("Exons + Introns")
```

### GABAergic

```{r}
cluster_num <- 3

markers <- FindMarkers(
  seurat_introns, ident.1 = paste(cluster_num, ident1, sep = "_"), ident.2 = paste(cluster_num, ident2, sep = "_"),
  only.pos = FALSE, verbose = FALSE
)
markers <- markers[markers$p_val_adj < 0.05, ]
datatable_download_exp(markers)

seurat_sub <- subset(
  seurat_introns, idents = c(paste(cluster_num, ident1, sep = "_"), paste(cluster_num, ident2, sep = "_"))
)
seurat_sub$seurat_clusters_sample_names <- sub(cluster_num, "GABAergic", seurat_sub$seurat_clusters_sample)
VlnPlot(seurat_sub, features = rownames(markers)[1:6], group.by = "seurat_clusters_sample_names") +
  plot_annotation("Exons + Introns")
```

## NLGF P301S+3 vs. MAPT KI

```{r}
ident1 <- "NLGF_P301S+3"
ident2 <- "MAPT_KI"

seurat_introns@active.ident <- factor(seurat_introns$sample)
markers <- FindMarkers(seurat_introns, ident.1 = ident1, ident.2 = ident2, only.pos = FALSE, verbose = FALSE)
markers <- markers[markers$p_val_adj < 0.05, ]
datatable_download_exp(markers)

seurat_sub <- subset(seurat_introns, idents = c(ident1, ident2))
VlnPlot(seurat_sub, features = rownames(markers)[1:6], group.by = "sample") + plot_annotation("Exons + Introns")

seurat_introns@active.ident <- factor(seurat_introns$seurat_clusters_sample)
```

### Oligodendrocytes

```{r}
cluster_num <- 2

markers <- FindMarkers(
  seurat_introns, ident.1 = paste(cluster_num, ident1, sep = "_"), ident.2 = paste(cluster_num, ident2, sep = "_"),
  only.pos = FALSE, verbose = FALSE
)
markers <- markers[markers$p_val_adj < 0.05, ]
datatable_download_exp(markers)

seurat_sub <- subset(
  seurat_introns, idents = c(paste(cluster_num, ident1, sep = "_"), paste(cluster_num, ident2, sep = "_"))
)
seurat_sub$seurat_clusters_sample_names <- sub(cluster_num, "Oligodendrocytes", seurat_sub$seurat_clusters_sample)
VlnPlot(seurat_sub, features = rownames(markers)[1:6], group.by = "seurat_clusters_sample_names") +
  plot_annotation("Exons + Introns")
```

### Microglia

```{r}
cluster_num <- 11

markers <- FindMarkers(
  seurat_introns, ident.1 = paste(cluster_num, ident1, sep = "_"), ident.2 = paste(cluster_num, ident2, sep = "_"),
  only.pos = FALSE, verbose = FALSE
)
markers <- markers[markers$p_val_adj < 0.05, ]
datatable_download_exp(markers)

seurat_sub <- subset(
  seurat_introns, idents = c(paste(cluster_num, ident1, sep = "_"), paste(cluster_num, ident2, sep = "_"))
)
seurat_sub$seurat_clusters_sample_names <- sub(cluster_num, "Microglia", seurat_sub$seurat_clusters_sample)
VlnPlot(seurat_sub, features = rownames(markers)[1:6], group.by = "seurat_clusters_sample_names") +
  plot_annotation("Exons + Introns")
```

### Astrocytes

```{r}
cluster_num <- 1

markers <- FindMarkers(
  seurat_introns, ident.1 = paste(cluster_num, ident1, sep = "_"), ident.2 = paste(cluster_num, ident2, sep = "_"),
  only.pos = FALSE, verbose = FALSE
)
markers <- markers[markers$p_val_adj < 0.05, ]
datatable_download_exp(markers)

seurat_sub <- subset(
  seurat_introns, idents = c(paste(cluster_num, ident1, sep = "_"), paste(cluster_num, ident2, sep = "_"))
)
seurat_sub$seurat_clusters_sample_names <- sub(cluster_num, "Astrocytes", seurat_sub$seurat_clusters_sample)
VlnPlot(seurat_sub, features = rownames(markers)[1:6], group.by = "seurat_clusters_sample_names") +
  plot_annotation("Exons + Introns")
```

### Cortical Glutamatergic

```{r}
cluster_num <- 0

markers <- FindMarkers(
  seurat_introns, ident.1 = paste(cluster_num, ident1, sep = "_"), ident.2 = paste(cluster_num, ident2, sep = "_"),
  only.pos = FALSE, verbose = FALSE
)
markers <- markers[markers$p_val_adj < 0.05, ]
datatable_download_exp(markers)

seurat_sub <- subset(
  seurat_introns, idents = c(paste(cluster_num, ident1, sep = "_"), paste(cluster_num, ident2, sep = "_"))
)
seurat_sub$seurat_clusters_sample_names <- sub(cluster_num, "Ctx_Glutamatergic", seurat_sub$seurat_clusters_sample)
VlnPlot(seurat_sub, features = rownames(markers)[1:6], group.by = "seurat_clusters_sample_names") +
  plot_annotation("Exons + Introns")
```

### GABAergic

```{r}
cluster_num <- 3

markers <- FindMarkers(
  seurat_introns, ident.1 = paste(cluster_num, ident1, sep = "_"), ident.2 = paste(cluster_num, ident2, sep = "_"),
  only.pos = FALSE, verbose = FALSE
)
markers <- markers[markers$p_val_adj < 0.05, ]
datatable_download_exp(markers)

seurat_sub <- subset(
  seurat_introns, idents = c(paste(cluster_num, ident1, sep = "_"), paste(cluster_num, ident2, sep = "_"))
)
seurat_sub$seurat_clusters_sample_names <- sub(cluster_num, "GABAergic", seurat_sub$seurat_clusters_sample)
VlnPlot(seurat_sub, features = rownames(markers)[1:6], group.by = "seurat_clusters_sample_names") +
  plot_annotation("Exons + Introns")
```

# References

This is the concluding section of the document, where we output the `sessionInfo`, and create a bibliography for works cited.

```{r}
sessionInfo()
```
