---
title: "NLGF P301S+3 - 01 Prep - No Introns"
author:
  - name: "Emir Turkes"
date: '`r strftime(Sys.time(), format = "%B %d, %Y")`'
link-citations: true
output:
  html_document:
    code_folding: hide
    number_sections: true
    theme: lumen
    highlight: haddock
    toc: true
    toc_depth: 2
    toc_float:
      collapsed: false
      smooth_scroll: false
knit: (function(inputFile, encoding) {
  rmarkdown::render(
    inputFile, encoding = encoding,
    output_file = file.path(
      "..", "..", "..", "results",
      unlist(strsplit(getwd(), "/"))[6], unlist(strsplit(getwd(), "/"))[7], "NLGF_P301S+3_01_prep_no_introns.html"
  )}))
---

<style type="text/css">
body {font-size: 16px;}
h1.title {font-size: 35px;}
h1 {font-size: 24px;}
h2 {font-size: 22px;}
.toc-content {padding-left: 0px; padding-right: 0px;}
div.tocify {width: 100%;}
.tocify-subheader .tocify-item {font-size: 0.95em; padding-left: 25px; text-indent: 0;}
div.main-container {max-width: none; width: 100%;}
</style>

*This file is a part of [tau-mutant-omics-analysis](https://github.com/eturkes/tau-mutant-omics-analysis).*

The table of contents in the top left is clickable and can be used to quickly navigate the document.
To toggle the visibility of code, use the `CODE` toggles at the top right of chunks.
The toggle at the start of the document controls the visibility of all chunks.

```{r}
#    This file is part of tau-mutant-omics-analysis.
#    Copyright (C) 2023  Emir Turkes, Naoto Watamura, UK DRI at UCL
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#    Emir Turkes can be contacted at emir.turkes@eturkes.com

# Load required packages, suppressing startup messages.
# -----------------------------------------------------
library(conflicted)
conflicts_prefer(matrixStats::rowMedians, .quiet = TRUE)
packages <- c(
  "SingleCellExperiment", "scuttle", "DT", "Seurat", "parallelly", "ggplot2", "dplyr", "ggrepel", "viridis",
  "DoubletFinder"
)
invisible(suppressPackageStartupMessages(lapply(packages, FUN = library, character.only = TRUE)))
# -----------------------------------------------------

# Define global settings.
# -----------------------
knitr::opts_chunk$set(fig.width = 10, fig.height = 7, dpi = 300)
# -----------------------

# Define functions.
# -----------------
source(file.path("..", "..", "utils.R"))
# -----------------

# Useful variables.
# -----------------
analysis_no <- 1
data_name <- unlist(strsplit(getwd(), "/"))[7] # Name of dataset derived from path.

data_dir <- file.path(
  "..", "..", "..", "data", "tau-mutant-omics-analysis",
  "low-coverage", "03-Cellranger_mapping", "Cellranger_output_without_introns", "NaotoS4_no_introns", "outs"
)
cache_dir <- file.path("..", "..", "..", "cache", "low-coverage", data_name, paste0("0", analysis_no, "-no-introns"))
if (!dir.exists(cache_dir)) {
  dir.create(cache_dir, recursive = TRUE)
}

protocol <- c("mouse", "droplet", "single-nuc", "umis") # See `cluster_pipeline` in `utils.R`.
vars_to_regress <- NULL # See `cluster_pipeline` in `utils.R`.
parallel_override <- 0 # See `parallel_plan` in `utils.R`.
# -----------------

# Metadata to plot after dimensionality reduction and clustering.
# Values in list can include "no_legend and/or "no_label" to exclude those.
# -------------------------------------------------------------------------
metadata_to_plot <- vector("list", length = 1)
names(metadata_to_plot) <- "seurat_clusters"
# -------------------------------------------------------------------------
```

# Cleaning

We first read in the data and form a SingleCellExperiment object.

```{r}
data <- Read10X(file.path(data_dir, "filtered_feature_bc_matrix"))

sce <- SingleCellExperiment(list(counts = data))
rm(data)

batch <- NULL # Specify if there are batches for QC sections.

sce
```

# QC

Then, we add cell QC metric metadata to the SCE object.

```{r}
mito <- grep("^MT-", rownames(sce), ignore.case = TRUE)

if (length(mito) != 0) {
  sce <- addPerCellQC(sce, list(mito = mito))
  if (is.null(batch)) {
    remove <- quickPerCellQC(colData(sce), percent_subsets = "subsets_mito_percent")
  } else {
    remove <- quickPerCellQC(colData(sce), percent_subsets = "subsets_mito_percent", batch = batch)
  }

  if (protocol[3] == "single-nuc") {
    cat(paste0("Removed ", length(mito), " mitochondrial genes:"), rownames(sce)[mito])
    sce <- sce[-mito, ]
    dim(sce)
  } else {
    cat(paste0("Found but did not remove ", length(mito), " mitochondrial genes:"), rownames(sce)[mito])
  }

} else {
  sce <- addPerCellQC(sce)
  if (is.null(batch)) {
    remove <- quickPerCellQC(colData(sce))
  } else {
    remove <- quickPerCellQC(colData(sce), batch = batch)
  }
}

names(colData(sce))[(length(names(colData(sce))) - 5):length(names(colData(sce)))]
```

We use adaptive thresholds to remove cells that are outliers by more than 3 MADs.

```{r}
sce$discard <- remove$discard
datatable_download(t(colSums(as.matrix(remove))))

sce <- sce[ , !sce$discard]
dim(sce)
```

Finally we add feature QC metrics and remove features not expressed in any cell.

```{r, fig.height = 5}
sce <- addPerFeatureQC(sce)
names(rowData(sce))
dim(sce)

par(mfrow = c(1, 3), mar = c(5, 4, 1, 1))
hist(
 log10(rowData(sce)$mean + 1e-6), col = "grey80",  main = "",
 breaks = 40, xlab = "log10(Mean Counts Per Gene + 1e-6)")
hist(
 log10((rowData(sce)$detected * dim(sce)[2]) + 1), col = "grey80", main = "",
 breaks = 40, xlab = "log10(Number of Cells Expressing Gene + 1)")
plot(
 log10(rowData(sce)$mean + 1e-6), pch = 16,
 col = rgb(0, 0, 0, 0.4), log10((rowData(sce)$detected * dim(sce)[2]) + 1),
 xlab = "log10(Mean Counts Per Gene + 1e-6)", ylab = "log10(Number of Cells Expressing Gene + 1)")

sce <- sce[rowSums(counts(sce) > 0) > 0, ]
rowData(sce) <- NULL
sce <- addPerFeatureQC(sce)
dim(sce)

par(mfrow = c(1, 3), mar = c(5, 4, 1, 1))
hist(
 log10(rowData(sce)$mean + 1e-6), col = "grey80",  main = "",
 breaks = 40, xlab = "log10(Mean Counts Per Gene + 1e-6)")
hist(
 log10((rowData(sce)$detected * dim(sce)[2]) + 1), col = "grey80", main = "",
 breaks = 40, xlab = "log10(Number of Cells Expressing Gene + 1)")
plot(
 log10(rowData(sce)$mean + 1e-6), pch = 16,
 col = rgb(0, 0, 0, 0.4), log10((rowData(sce)$detected * dim(sce)[2]) + 1),
 xlab = "log10(Mean Counts Per Gene + 1e-6)", ylab = "log10(Number of Cells Expressing Gene + 1)")
```

# Cluster Pipeline

We run a pipeline that applies normalization and scaling, dimensionality reduction, and clustering.

```{r}
# Remove redundant QC metrics.
# ----------------------------
if (length(mito) != 0) {
  colData(sce)[(length(names(colData(sce))) - 1):length(names(colData(sce)))] <- NULL
  colData(sce)[(length(names(colData(sce))) - 4):(length(names(colData(sce))) - 2)] <- NULL
} else {
  colData(sce)[(length(names(colData(sce))) - 2):length(names(colData(sce)))] <- NULL
}
rowData(sce) <- NULL
# ----------------------------

sub_name <- "all"
seurat <- as.Seurat(sce, data = NULL)
rm(sce)
seurat <- RenameAssays(seurat, originalexp = "RNA")
seurat <- cluster_pipeline(
  seurat, cache_dir = cache_dir, sub_name = sub_name, protocol = protocol,
  vars_to_regress = vars_to_regress, parallel_override = parallel_override, cc = FALSE, res_divider = 100000
)
seurat

for (i in 1:length(metadata_to_plot)) {
  if ("no_legend" %in% metadata_to_plot[[i]] && "no_label" %in% metadata_to_plot[[i]]) {
    print(
      red_dim_plot(seurat, x = "umap1", y = "umap2", color = names(metadata_to_plot)[i]) +
        NoLegend() +
        ggtitle(names(metadata_to_plot)[i])
    )
  } else if ("no_legend" %in% metadata_to_plot[[i]]) {
    print(
      red_dim_plot(seurat, x = "umap1", y = "umap2", color = names(metadata_to_plot)[i], type = "cat") +
        NoLegend() +
        ggtitle(names(metadata_to_plot)[i])
    )
  } else if ("no_label" %in% metadata_to_plot[[i]]) {
    print(
      red_dim_plot(seurat, x = "umap1", y = "umap2", color = names(metadata_to_plot)[i]) +
        ggtitle(names(metadata_to_plot)[i])
    )
  } else {
    print(
      red_dim_plot(seurat, x = "umap1", y = "umap2", color = names(metadata_to_plot)[i], type = "cat") +
        ggtitle(names(metadata_to_plot)[i])
    )
  }
}

if (protocol[4] == "umis") {
  print(red_dim_plot(seurat, x = "umap1", y = "umap2", color = "nFeature_SCT", type = "cont") + ggtitle("nFeature_SCT"))
  print(red_dim_plot(seurat, x = "umap1", y = "umap2", color = "nCount_SCT", type = "cont") + ggtitle("nCount_SCT"))
} else if (protocol[4] == "reads") {
  print(red_dim_plot(seurat, x = "umap1", y = "umap2", color = "nFeature_RNA", type = "cont") + ggtitle("nFeature_RNA"))
  print(red_dim_plot(seurat, x = "umap1", y = "umap2", color = "nCount_RNA", type = "cont") + ggtitle("nCount_RNA"))
}

if (length(mito) != 0) {
  red_dim_plot(seurat, x = "umap1", y = "umap2", color = "subsets_mito_percent", type = "cont") +
    ggtitle("subsets_mito_percent")
}
```

# References

This is the concluding section of the document, where we output the `sessionInfo`, and create a bibliography for works cited.

```{r}
sessionInfo()
```
